File: crawler/src/main/java/org/beckn/crawler/crawl/Fetcher.java
package org.beckn.crawler.crawl;

import org.beckn.crawler.http.CrawlerHttpClient;
import org.beckn.crawler.util.DigestUtil;
import org.springframework.stereotype.Component;

import java.io.IOException;

/**
 * Step 5 (design doc §5.4): GET a catalog part and verify its bytes against the digest the
 * index published. A mismatch is a hard reject — never index unverified bytes (§5.7).
 */
@Component
public class Fetcher {

    /** Raised when a part's bytes don't match its announced digest. */
    public static class DigestMismatchException extends Exception {
        public DigestMismatchException(String message) { super(message); }
    }

    private final CrawlerHttpClient http;

    public Fetcher(CrawlerHttpClient http) {
        this.http = http;
    }

    /** Returns the verified raw bytes of the part (safe to push). */
    public byte[] fetchVerified(String partUrl, String expectedDigest)
            throws IOException, InterruptedException, DigestMismatchException {
        CrawlerHttpClient.Response resp = http.get(partUrl);
        if (resp.status() != 200) {
            throw new IOException("part GET " + partUrl + " returned HTTP " + resp.status());
        }
        if (!DigestUtil.matches(resp.body(), expectedDigest)) {
            throw new DigestMismatchException("expected " + expectedDigest
                    + " got " + DigestUtil.sha256(resp.body()));
        }
        return resp.body();
    }
}

File: crawler-go/internal/crawl/fetcher.go
package crawl

import (
	"context"
	"fmt"

	"github.com/beckn/beckn-discovr/crawler-go/internal/digest"
	"github.com/beckn/beckn-discovr/crawler-go/internal/httpclient"
)

// DigestMismatchError is raised when a part's bytes don't match its announced digest — a hard
// reject (never index unverified bytes, design doc §5.7). Ported from Fetcher.DigestMismatchException.
type DigestMismatchError struct{ msg string }

func (e *DigestMismatchError) Error() string { return e.msg }

// Fetcher GETs a catalog part and verifies its bytes against the digest the index published.
type Fetcher struct {
	http *httpclient.Client
}

// NewFetcher builds a Fetcher over the shared HTTP client.
func NewFetcher(http *httpclient.Client) *Fetcher { return &Fetcher{http: http} }

// FetchVerified returns the verified raw bytes of the part (safe to push).
func (f *Fetcher) FetchVerified(ctx context.Context, partURL, expectedDigest string) ([]byte, error) {
	resp, err := f.http.Get(ctx, partURL)
	if err != nil {
		return nil, err
	}
	if resp.Status != 200 {
		return nil, fmt.Errorf("part GET %s returned HTTP %d", partURL, resp.Status)
	}
	if !digest.Matches(resp.Body, expectedDigest) {
		return nil, &DigestMismatchError{
			msg: fmt.Sprintf("expected %s got %s", expectedDigest, digest.SHA256(resp.Body)),
		}
	}
	return resp.Body, nil
}
